provider/datagen/src/bin/datagen/mod.rs

```diff
@@ -86,23 +86,52 @@ fn main() -> eyre::Result<()> {
         config::KeyInclude::Explicit(set) => driver.with_keys(set),
         config::KeyInclude::ForBinary(path) => driver.with_keys(icu_datagen::keys_from_bin(path)?),
     };
-    driver = driver.with_fallback_mode(config.fallback);
-    driver = driver.with_additional_collations(config.additional_collations);
-    driver = match config.locales {
-        config::LocaleInclude::All => driver.with_all_locales(),
-        config::LocaleInclude::None => driver.with_locales([]),
-        config::LocaleInclude::Explicit(set) => driver.with_locales(set),
-        config::LocaleInclude::CldrSet(levels) => {
-            driver.with_locales(provider.locales_for_coverage_levels(levels.iter().copied())?)
-        }
-        config::LocaleInclude::Recommended => {
-            driver.with_locales(provider.locales_for_coverage_levels([
-                CoverageLevel::Modern,
-                CoverageLevel::Moderate,
-                CoverageLevel::Basic,
-            ])?)
-        }
+    enum LanguageIdentifiersOrLocaleFamilies {
+        LanguageIdentifiers(Vec<LanguageIdentifier>),
+        LocaleFamilies(Vec<LocaleFamily>),
+    }
+    use LanguageIdentifiersOrLocaleFamilies::*;
+    let locales_intermediate: LanguageIdentifiersOrLocaleFamilies = match config.locales {
```

Comment:
suggestion:

resolve `::CldrSet`, `::Recommended`, and `::None` to `::Explicit` and handle the `::All` and `::Explicit` cases explicitly below. This avoids the `locales_intermediate`.

Reply:
This is internal. I may rewrite slightly once LocaleFamily is done. However, I prefer to merge them into a custom internal enum than merge them into the config enum and have unreachable branches.